File: backend/WorldBeauty/Empresa.java
package WorldBeauty;

import java.util.ArrayList;
import java.util.List;

public class Empresa {
    private static List<Cliente> clientes = new ArrayList<Cliente>();
    private static List<Produto> produtos = new ArrayList<Produto>();
    private static List<Servico> servicos = new ArrayList<Servico>();

    // private static LoadData() {
    // // clientes = JSON.parse(localStorage.getItem("Clientes") as string) ||
    // [];
    // // produtos = JSON.parse(localStorage.getItem("Produtos") as string) ||
    // [];
    // // servicos = JSON.parse(localStorage.getItem("Servicos") as string) ||
    // [];
    // clientes = [];
    // const clientes = JSON.parse(localStorage.getItem("Clientes") as string) ||
    // [];
    // clientes.forEach((cliente any) => {
    // clientes.push(
    // new Cliente(
    // cliente.nome,
    // cliente.nomeSocial,
    // cliente.genero,
    // cliente.CPF,
    // cliente.RGs,
    // cliente.dataCadastro,
    // cliente.telefones,
    // cliente.produtosConsumidos,
    // cliente.servicosConsumidos
    // )
    // );
    // });

    // produtos = [];
    // const produtos = JSON.parse(localStorage.getItem("Produtos") as string) ||
    // [];
    // produtos.forEach((produto any) => {
    // produtos.push(
    // new Produto(produto.nome, produto.valor)
    // );
    // });

    // servicos = [];
    // const servicos = JSON.parse(localStorage.getItem("Servicos") as string) ||
    // [];
    // servicos.forEach((servico any) => {
    // servicos.push(
    // new Servico(servico.nome, servico.valor)
    // );
    // });
    // }

    // public static SaveData() {
    // localStorage.setItem("Clientes", JSON.stringify(clientes));
    // localStorage.setItem("Produtos", JSON.stringify(produtos));
    // localStorage.setItem("Servicos", JSON.stringify(servicos));
    // }

    public static void AdicionarCliente(Cliente cliente) {
        clientes.add(cliente);
    }

    public static void AdicionarProduto(Produto produto) {
        produtos.add(produto);
    }

    public static void AdicionarServico(Servico servico) {
        servicos.add(servico);
    }

    public static void DeletarCliente(CPF cpf) {
        clientes.removeIf(
                c -> (c.GetCPF().GetValor() == cpf.GetValor() && c.GetCPF().GetDataEmissao() == cpf.GetDataEmissao()));
    }

    public static void DeletarProduto(String nome) {
        produtos.removeIf(p -> (p.GetNome() == nome));
    }

    public static void DeletarServico(String nome) {
        servicos.removeIf(s -> (s.GetNome() == nome));
    }

    public static void AtualizarCliente(CPF cpf, String nome, String nomeSocial, String genero, List<RG> RGs,
            List<Telefone> telefones) {
        for (int i = 0; i < clientes.size(); i++) {
            if (clientes.get(i).GetCPF().GetValor() == cpf.GetValor()
                    && clientes.get(i).GetCPF().GetDataEmissao() == cpf.GetDataEmissao()) {
                clientes.get(i).SetNome(nome);
                clientes.get(i).SetNomeSocial(nomeSocial);
                clientes.get(i).SetGenero(genero);
                clientes.get(i).SetRGs(RGs);
                clientes.get(i).SetTelefones(telefones);
                break;
            }
        }
    }

    public static void AtualizarProduto(String nome, float valor) {
        for (int i = 0; i < produtos.size(); i++) {
            if (produtos.get(i).GetNome() == nome) {
                produtos.get(i).SetValor(valor);
                break;
            }
        }
    }

    public static void AtualizarServico(String nome, float valor) {
        for (int i = 0; i < servicos.size(); i++) {
            if (servicos.get(i).GetNome() == nome) {
                servicos.get(i).SetValor(valor);
                break;
            }
        }
    }

    public static List<Cliente> GetClientes() {
        return clientes;
    }

    public static List<Produto> GetProdutos() {
        return produtos;
    }

    public static List<Servico> GetServicos() {
        return servicos;
    }

    public static Cliente[] GetTopXClientesConsumoProdutoQtd(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> (c1.GetProdutosConsumidos().size() - c2.GetProdutosConsumidos().size()));

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(i);
        }

        return retClientes;
    }

    public static Cliente[] GetTopXClientesConsumoProdutoVal(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> {
            List<Float> valsA = new ArrayList<Float>();
            c1.GetProdutosConsumidos().forEach(p -> valsA.add(p.GetValor()));

            List<Float> valsB = new ArrayList<Float>();
            c2.GetProdutosConsumidos().forEach(p -> valsB.add(p.GetValor()));

            float valA = 0;
            for (float f : valsA) valA += f;
            float valB = 0;
            for (float f : valsB) valB += f;
            return (int) (valA - valB);
        });

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(i);
        }

        return retClientes;
    }

    public static Cliente[] GetSubXClientesConsumoProdutoQtd(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> (c1.GetProdutosConsumidos().size() - c2.GetProdutosConsumidos().size()));

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(clientes.size() - (i + 1));
        }

        return retClientes;
    }

    public static Cliente[] GetSubXClientesConsumoProdutoVal(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> {
            List<Float> valsA = new ArrayList<Float>();
            c1.GetProdutosConsumidos().forEach(p -> valsA.add(p.GetValor()));

            List<Float> valsB = new ArrayList<Float>();
            c2.GetProdutosConsumidos().forEach(p -> valsB.add(p.GetValor()));

            float valA = 0;
            for (float f : valsA) valA += f;
            float valB = 0;
            for (float f : valsB) valB += f;
            return (int) (valA - valB);
        });

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(clientes.size() - (i + 1));
        }

        return retClientes;
    }

    public static Cliente[] GetTopXClientesConsumoServicoQtd(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> (c1.GetServicosConsumidos().size() - c2.GetServicosConsumidos().size()));

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(i);
        }

        return retClientes;
    }

    public static Cliente[] GetTopXClientesConsumoServicoVal(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> (c1.GetServicosConsumidos().size() - c2.GetServicosConsumidos().size()));

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(clientes.size() - (i + 1));
        }

        return retClientes;
    }

    public static Cliente[] GetSubXClientesConsumoServicoQtd(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> (c1.GetServicosConsumidos().size() - c2.GetServicosConsumidos().size()));

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(clientes.size() - (i + 1));
        }

        return retClientes;
    }

    public static Cliente[] GetSubXClientesConsumoServicoVal(int qtd) {
        Cliente[] retClientes = new Cliente[qtd];

        clientes.sort((c1, c2) -> {
            List<Float> valsA = new ArrayList<Float>();
            c1.GetProdutosConsumidos().forEach(p -> valsA.add(p.GetValor()));

            List<Float> valsB = new ArrayList<Float>();
            c2.GetProdutosConsumidos().forEach(p -> valsB.add(p.GetValor()));

            float valA = 0;
            for (float f : valsA) valA += f;
            float valB = 0;
            for (float f : valsB) valB += f;
            return (int) (valA - valB);
        });

        for (int i = 0; i < qtd; i++) {
            retClientes[i] = clientes.get(clientes.size() - (i + 1));
        }

        return retClientes;
    }

    public static Cliente[] ClientesDoGenero(String genero) {
        List<Cliente> retClientes = new ArrayList<Cliente>(0);
        clientes.forEach(cli -> {
            if (cli.GetGenero() == genero.toUpperCase())
                retClientes.add(cli);
        });
        return (Cliente[]) retClientes.toArray();
    }
}

File: world-beauty/src/WB-ATV1/src/Business/Empresa.ts
import Cliente from "../Models/Cliente";
import Produto from "../Models/Produto";
import RG from "../Models/RG";
import Servico from "../Models/Servico";
import Telefone from "../Models/Telefone";

export default class Empresa {
    private static clientes: Array<Cliente> = new Array<Cliente>(0);
    private static produtos: Array<Produto> = new Array<Produto>(0);
    private static servicos: Array<Servico> = new Array<Servico>(0);

    private static LoadData() {
        // this.clientes = JSON.parse(localStorage.getItem("Clientes") as string) || [];
        // this.produtos = JSON.parse(localStorage.getItem("Produtos") as string) || [];
        // this.servicos = JSON.parse(localStorage.getItem("Servicos") as string) || [];
        this.clientes = [];
        const clientes = JSON.parse(localStorage.getItem("Clientes") as string) || [];
        clientes.forEach((cliente: any) => {
            this.clientes.push(
                new Cliente(
                    cliente.nome,
                    cliente.nomeSocial,
                    cliente.genero,
                    cliente.CPF,
                    cliente.RGs,
                    cliente.dataCadastro,
                    cliente.telefones,
                    cliente.produtosConsumidos,
                    cliente.servicosConsumidos
                )
            );
        });

        this.produtos = [];
        const produtos = JSON.parse(localStorage.getItem("Produtos") as string) || [];
        produtos.forEach((produto: any) => {
            this.produtos.push(
                new Produto(produto.nome, produto.valor)
            );
        });

        this.servicos = [];
        const servicos = JSON.parse(localStorage.getItem("Servicos") as string) || [];
        servicos.forEach((servico: any) => {
            this.servicos.push(
                new Servico(servico.nome, servico.valor)
            );
        });
    }

    public static SaveData() {
        localStorage.setItem("Clientes", JSON.stringify(this.clientes));
        localStorage.setItem("Produtos", JSON.stringify(this.produtos));
        localStorage.setItem("Servicos", JSON.stringify(this.servicos));
    }

    public static AdicionarCliente(cliente: Cliente) {
        this.LoadData();
        this.clientes.push(cliente);
        this.SaveData();
    }
    public static AdicionarProduto(produto: Produto) {
        this.LoadData();
        this.produtos.push(produto);
        this.SaveData();
    }
    public static AdicionarServico(servico: Servico) {
        this.LoadData();
        this.servicos.push(servico);
        this.SaveData();
    }

    public static DeletarCliente(cpf: any) {
        this.LoadData();
        this.clientes = this.clientes.filter((cliente) => {
            return ((cliente.GetCPF as any).valor != cpf.valor) && ((cliente.GetCPF as any).dataEmissao != cpf.dataEmissao);
        });
        this.SaveData();
    }
    public static DeletarProduto(nome: string) {
        this.LoadData();
        this.produtos = this.produtos.filter((produto) => { return produto.GetNome != nome });
        this.SaveData();
    }
    public static DeletarServico(nome: string) {
        this.LoadData();
        this.servicos = this.servicos.filter((servico) => { return servico.GetNome != nome });
        this.SaveData();
    }

    public static AtualizarCliente(cpf: any, nome: string, nomeSocial: string, genero: string, RGs: Array<RG>, telefones: Array<Telefone>) {
        this.LoadData();
        const cliente = this.clientes.filter((cliente) => (cliente.GetCPF as any).valor === cpf.valor && (cliente.GetCPF as any).dataEmissao === cpf.dataEmissao)[0];
        if (cliente) {
            cliente.SetNome = nome;
            cliente.SetNomeSocial = nomeSocial;
            cliente.SetGenero = genero;
            cliente.SetRGs = RGs;
            cliente.SetTelefones = telefones;
        }
        this.SaveData();
    }
    public static AtualizarProduto(nome: string, valor: number) {
        this.LoadData();
        const produto = this.produtos.find((produto) => produto.GetNome === nome);
        if (produto) {
            produto.SetValor = valor;
        }
        this.SaveData();
    }
    public static AtualizarServico(nome: string, valor: number) {
        this.LoadData();
        const servico = this.produtos.find((servico) => servico.GetNome === nome);
        if (servico) {
            servico.SetValor = valor;
        }
        this.SaveData();
    }

    public static get GetClientes() {
        this.LoadData();
        return this.clientes;
    }

    public static get GetProdutos() {
        this.LoadData();
        return this.produtos;
    }

    public static get GetServicos() {
        this.LoadData();
        return this.servicos;
    }

    public static GetTopXClientesConsumoProdutoQtd(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                return a.GetProdutosConsumidos.length - b.GetProdutosConsumidos.length;
            }
        ).slice(0, qtd);
    }
    public static GetTopXClientesConsumoProdutoVal(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                let valA = 0;
                a.GetProdutosConsumidos.forEach(
                    function (produto: Produto) {
                        valA += produto.GetValor;
                    }
                )

                let valB = 0;
                b.GetProdutosConsumidos.forEach(
                    function (produto: Produto) {
                        valB += produto.GetValor;
                    }
                )

                return valA - valB;
            }
        ).slice(0, qtd);
    }
    public static GetSubXClientesConsumoProdutoQtd(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                return b.GetProdutosConsumidos.length - a.GetProdutosConsumidos.length;
            }
        ).slice(0, qtd);
    }
    public static GetSubXClientesConsumoProdutoVal(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                let valA = 0;
                a.GetProdutosConsumidos.forEach(
                    function (produto: Produto) {
                        valA += produto.GetValor;
                    }
                )

                let valB = 0;
                b.GetProdutosConsumidos.forEach(
                    function (servico: Produto) {
                        valB += servico.GetValor;
                    }
                )

                return valB - valA;
            }
        ).slice(0, qtd);
    }

    public static GetTopXClientesConsumoServicoQtd(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                return a.GetServicosConsumidos.length - b.GetServicosConsumidos.length;
            }
        ).slice(0, qtd);
    }
    public static GetTopXClientesConsumoServicoVal(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                let valA = 0;
                a.GetServicosConsumidos.forEach(
                    function (servico: Servico) {
                        valA += servico.GetValor;
                    }
                )

                let valB = 0;
                b.GetServicosConsumidos.forEach(
                    function (servico: Servico) {
                        valB += servico.GetValor;
                    }
                )

                return valA - valB;
            }
        ).slice(0, qtd);
    }
    public static GetSubXClientesConsumoServicoQtd(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                return b.GetServicosConsumidos.length - a.GetServicosConsumidos.length;
            }
        ).slice(0, qtd);
    }
    public static GetSubXClientesConsumoServicoVal(qtd: number): Array<Cliente> {
        this.LoadData();
        return this.clientes.sort(
            function (a: Cliente, b: Cliente): number {
                let valA = 0;
                a.GetServicosConsumidos.forEach(
                    function (servico: Servico) {
                        valA += servico.GetValor;
                    }
                )

                let valB = 0;
                b.GetServicosConsumidos.forEach(
                    function (servico: Servico) {
                        valB += servico.GetValor;
                    }
                )

                return valB - valA;
            }
        ).slice(0, qtd);
    }

    public static ClientesDoGenero(genero: string): Array<Cliente> {
        this.LoadData();
        return this.clientes.filter(cli => cli.GetGenero === genero.toUpperCase());
    }
}